### DISCUSSION TITLE
feature request: .data() method for nb::bytes

### DISCUSSION THREAD

**Discussion by jeffective** [wjakob/nanobind#569] [Q&A]
Can we have a .data() method for the nb::bytes object to return a pointer to the underlying data?

This could be helpful when using c libraries that pass around size and pointers to the starts of buffers.

**Comment by nurpax** [comment]
I was recently dealing with `nb::bytes` too and noticed that the current accessor _does_ work work, it's just that it looks like it's only intended for C strings.

So if you have a bytes object in C++, use `bytes_obj.size()` for size and `bytes_obj.c_str()` for the data pointer.  You'll get the whole data buffer even if it contains char code 0s.

**Reply by wjakob** [reply]
Could you make a PR that adds this and also documents the new method?

**Reply by nurpax** [reply]
I can send a PR if I can figure out how to build for nanobind development and run the test suite.. :)

**Reply by nurpax** [reply]
I tried with:

```
# in a venv, under a nanobind checkout
pip install .  # <-- this works fine
pip install pytest
pytest
```

But this errors out with:

```
================================================= test session starts =================================================
platform win32 -- Python 3.11.7, pytest-8.2.0, pluggy-1.5.0
rootdir: C:\Users\janne\dev\nanobind
configfile: pyproject.toml
testpaths: tests
collected 92 items / 16 errors

======================================================= ERRORS ========================================================
________________________________________ ERROR collecting tests/test_chrono.py ________________________________________
ImportError while importing test module 'C:\Users\janne\dev\nanobind\tests\test_chrono.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
..\..\python\Lib\importlib\__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests\test_chrono.py:3: in <module>
    import test_chrono_ext as m
E   ModuleNotFoundError: No module named 'test_chrono_ext'

...

=============================================== short test summary info ===============================================
ERROR tests/test_chrono.py
ERROR tests/test_classes.py
ERROR tests/test_enum.py
ERROR tests/test_eval.py
ERROR tests/test_exception.py
ERROR tests/test_functions.py
ERROR tests/test_holders.py
ERROR tests/test_inter_module.py
ERROR tests/test_intrusive.py
ERROR tests/test_issue.py
ERROR tests/test_make_iterator.py
ERROR tests/test_ndarray.py
ERROR tests/test_stl.py
ERROR tests/test_stl_bind_map.py
ERROR tests/test_stl_bind_vector.py
ERROR tests/test_typing.py
```

Not sure if this is something to do with me running on Windows.

**Reply by nurpax** [reply]
@wjakob Is this close to what you had in mind: https://github.com/wjakob/nanobind/pull/584?  If so, I can amend the PR to add docs and pass CI.  (Looks like they're not passing due to mismatched .pyi files for the test extension)
